plugins/dashboard/ws.go

```diff
@@ -0,0 +1,146 @@
+package dashboard
+
+import (
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/iotaledger/goshimmer/packages/shutdown"
+	"github.com/iotaledger/goshimmer/plugins/messagelayer"
+	"github.com/iotaledger/goshimmer/plugins/metrics"
+	"github.com/iotaledger/hive.go/daemon"
+	"github.com/iotaledger/hive.go/events"
+	"github.com/iotaledger/hive.go/workerpool"
+	"github.com/labstack/echo"
+)
+
+var (
+	// settings
+	wsSendWorkerCount     = 1
+	wsSendWorkerQueueSize = 250
+	wsSendWorkerPool      *workerpool.WorkerPool
+	webSocketWriteTimeout = time.Duration(3) * time.Second
+
+	// clients
+	wsClientsMu    sync.Mutex
+	wsClients      = make(map[uint64]*wsclient)
+	nextWsClientID uint64
+
+	// gorilla websocket layer
+	upgrader = websocket.Upgrader{
+		HandshakeTimeout:  webSocketWriteTimeout,
+		CheckOrigin:       func(r *http.Request) bool { return true },
+		EnableCompression: true,
+	}
+)
+
+// a websocket client with a channel for downstream messages.
+type wsclient struct {
+	// downstream message channel.
+	channel chan interface{}
+	// a channel which is closed when the websocket client is disconnected.
+	exit chan struct{}
+}
+
+func configureWebSocketWorkerPool() {
+	wsSendWorkerPool = workerpool.New(func(task workerpool.Task) {
+		broadcastWsMessage(&wsmsg{MsgTypeMPSMetric, task.Param(0).(uint64)})
+		broadcastWsMessage(&wsmsg{MsgTypeNodeStatus, currentNodeStatus()})
+		broadcastWsMessage(&wsmsg{MsgTypeNeighborMetric, neighborMetrics()})
+		broadcastWsMessage(&wsmsg{MsgTypeTipsMetric, messagelayer.TipSelector.TipCount()})
+		task.Return(nil)
+	}, workerpool.WorkerCount(wsSendWorkerCount), workerpool.QueueSize(wsSendWorkerQueueSize))
+}
+
+func runWebSocketStreams() {
+	updateStatus := events.NewClosure(func(mps uint64) {
+		wsSendWorkerPool.TrySubmit(mps)
+	})
+
+	daemon.BackgroundWorker("Dashboard[StatusUpdate]", func(shutdownSignal <-chan struct{}) {
+		metrics.Events.ReceivedMPSUpdated.Attach(updateStatus)
```

Comment:
If the Echo server stops due to an error (maybe the bind address is already used), we are still doing all these updates for no reason. Is there an easy fix for that? If not let's just leave it like this.

Reply:
I believe GoShimmer should shutdown in that case automatically but for that we'd need some new mechanism to do that. Perhaps we can expose `gracefulStop` from the `gracefulshutdown` plugin in order to make it shutdown properly.

Reply:
Currently, the only option we would have for this is to panic or exit, which is indeed not very nice.
Had a quick look at the `gracefulshutdown` plugin and it seems to be broken as `daemon.GetRunningBackgroundWorkers` is locked during shutdown. So, it does not do more than just calling `daemon.Shutdown()`.
Maybe we could just create a util function somewhere that calls daemon.Shutdown with and prints an error. That could be used for all the errors that should lead to a graceful shutdown...

Reply:
https://github.com/iotaledger/goshimmer/issues/385 to track this